### ISSUE TITLE
show system menu when right click titlebar

### ISSUE THREAD

**Issue by maz-1** [wangwenx190/framelesshelper#42]
What I mean is this menu:
![menu](https://user-images.githubusercontent.com/5277060/109829500-8ee19400-7c78-11eb-8496-bfc64332054c.png)

I use followng code to show this menu when left-click on titlebar icon
```
    QObject::connect(mainWindow,
                     &QMainWindow::windowIconChanged,
                     titleBarWidget.iconButton,
                     &QPushButton::setIcon);
    QObject::connect(titleBarWidget.iconButton,
    &QPushButton::clicked, mainWindow, [mainWindow] () {
#ifdef WIN32
        HWND hwnd = (HWND)QGuiApplication::platformNativeInterface()->nativeResourceForWindow(QByteArrayLiteral("handle"), mainWindow->windowHandle());
        //mainWindow->winId()
        HMENU systemMenu = ::GetSystemMenu(hwnd, FALSE);
        if (systemMenu) {
            // Open system menu in the left top corner of window.
            double pos_x = static_cast<double>(mainWindow->pos().x()) * mainWindow->devicePixelRatioF();
            double pos_y = static_cast<double>(mainWindow->pos().y() + mainWindow->menuWidget()->height()) * mainWindow->devicePixelRatioF();
            TrackPopupMenu(systemMenu, TPM_LEFTALIGN | TPM_LEFTBUTTON,
                           pos_x, pos_y, NULL,
                           hwnd, 0);
        }
#endif
    });
```

However I don't know how to show this when user right-click titlebar as well, any idea?

**Comment by maz-1** [comment]
All mouse events will be blocked for areas not added to ignored object list, so seems that I need to modify framelesshelper.cpp, which I have no idea where to get started.

**Comment by wangwenx190** [comment]
>
> I use followng code to show this menu when left-click on titlebar icon
>
Your code is correct in principle.

>
> I have no idea where to get started.
>
Well, last year I tried to add the feature you mentioned to this repo, I spend several months on it, but I failed in the end. The most critical thing is I can't get the context menu message in the native event filter, maybe Qt itself filtered it out. But actually it's achieveable, but it can't be done in the native event filter, you have to do it in the QWidget/QWindow's nativeEvent function. You can try it yourself. Related messages are `WM_NCRBUTTONUP`, `WM_CONTEXTMENU`, you just need to judge whether the mouse is in the title bar area first.

**Comment by wangwenx190** [comment]
>
> seems that I need to modify framelesshelper.cpp
>
No it's not needed. I've tried it long ago, it will not work. You have to override QWidget/QWindow's nativeEvent to do that.

**Comment by JulienMaille** [comment]
@maz-1 which includes are needed for this to work? I would like to add context menu to the qmainwindow example
Thank you

**Comment by wangwenx190** [comment]
> @maz-1 which includes are needed for this to work? I would like to add context menu to the qmainwindow example
> Thank you

The needed Win32 API is `TrackPopupMenu`

**Comment by wangwenx190** [comment]
@JulienMaille You need to do it in the WM_NCRBUTTONUP or WM_CONTEXTMENU

**Comment by JulienMaille** [comment]
commited :)

**Comment by maz-1** [comment]
> @maz-1 which includes are needed for this to work? I would like to add context menu to the qmainwindow example
> EDIT: I was able to use your code and show the menu, but clicking on items do not trigger anything (min/max/close)


My previous code does has some problem, you will need something like this:
```
static inline void displaySystemMenu(HWND hWnd, float pos_x, float pos_y)
{
    HMENU hMenu = ::GetSystemMenu(hWnd, FALSE);
    if (hMenu) {
        MENUITEMINFO mii;
        mii.cbSize = sizeof(MENUITEMINFO);
        mii.fMask = MIIM_STATE;
        mii.fType = 0;
        // update the options
        mii.fState = MF_ENABLED;
        SetMenuItemInfo(hMenu, SC_RESTORE, FALSE, &mii);
        SetMenuItemInfo(hMenu, SC_SIZE, FALSE, &mii);
        SetMenuItemInfo(hMenu, SC_MOVE, FALSE, &mii);
        SetMenuItemInfo(hMenu, SC_MAXIMIZE, FALSE, &mii);
        SetMenuItemInfo(hMenu, SC_MINIMIZE, FALSE, &mii);
        mii.fState = MF_GRAYED;
        WINDOWPLACEMENT wp;
        GetWindowPlacement(hWnd, &wp);
        switch (wp.showCmd) {
            case SW_SHOWMAXIMIZED:
                SetMenuItemInfo(hMenu, SC_SIZE, FALSE, &mii);
                SetMenuItemInfo(hMenu, SC_MOVE, FALSE, &mii);
                SetMenuItemInfo(hMenu, SC_MAXIMIZE, FALSE, &mii);
                SetMenuDefaultItem(hMenu, SC_CLOSE, FALSE);
                break;
            case SW_SHOWMINIMIZED:
                SetMenuItemInfo(hMenu, SC_MINIMIZE, FALSE, &mii);
                SetMenuDefaultItem(hMenu, SC_RESTORE, FALSE);
                break;
            case SW_SHOWNORMAL:
                SetMenuItemInfo(hMenu, SC_RESTORE, FALSE, &mii);
                SetMenuDefaultItem(hMenu, SC_CLOSE, FALSE);
                break;
        }
        LPARAM cmd = TrackPopupMenu(hMenu, (TPM_RIGHTBUTTON | TPM_NONOTIFY | TPM_RETURNCMD),
                                    pos_x, pos_y, NULL, hWnd, NULL);
        if (cmd) {
            PostMessage(hWnd, WM_SYSCOMMAND, cmd, 0);
        }
    }
}
```

**Comment by JulienMaille** [comment]
@wangwenx190 shouldn't we call `Utilities::showSystemMenu` from the same place we test the mouse and report `HTCAPTION`?

**Comment by wangwenx190** [comment]
> @wangwenx190 shouldn't we call `Utilities::showSystemMenu` from the same place we test the mouse and report `HTCAPTION`?

Normally this really should happen in the `WM_NCRBUTTONUP` block. `WM_NCHITTEST` will be triggered even if the mouse is hovering

**Comment by JulienMaille** [comment]
It seems my recent fix prevents this test to be true
https://github.com/wangwenx190/framelesshelper/blob/9904a5077ada12e791807cae1571161b3035c1fe/utilities_win32.cpp#L432

**Comment by wangwenx190** [comment]
Your fix doesn't seem to be related to this, need further investigate

**Comment by wangwenx190** [comment]
@JulienMaille  Yes indeed, your code is causing WM_NCRBUTTONUP being skipped, but I won't revert your code, I think the fix is more important than this issue, and this issue seem to can be solved by some other ways, need more investigation though.

**Comment by JulienMaille** [comment]
I think my code is a workaround. I currently use a different workaround that doesn't break menu. I could PR it it you want

**Comment by wangwenx190** [comment]
> I think my code is a workaround. I currently use a different workaround that doesn't break menu. I could PR it it you want

Sure, bro! I'm eager to know the better workaround! :)

**Comment by wangwenx190** [comment]
Implemented in 2.0
